Create new "splash-config" command

jsrc27 opened the pull request:
Create new `splash-config` command that can customize the Plymouth configuration file for our splash screen on Torizon OS. Also did some refactoring on the `splash` command so that it can work more smoothly with `splash-config` and any future features that are also related to splash-screen configuration.

lucas-akira reviewed (commented):
New feature looks good to me. One question: Should we add an option to reset the splash config back to the default one from the unpacked image, in a similar way we have the `kernel clear_custom_args` command?

jsrc27 commented:
> New feature looks good to me. One question: Should we add an option to reset the splash config back to the default one from the unpacked image, in a similar way we have the kernel clear_custom_args command?

I thought about it. But I think it's best to use unpacking the image as a more standard way to globally reset the image of any customization.

rborn-tx commented on tcbuilder/cli/union.py:
```diff
@@ -195,6 +195,12 @@ def union(changes_dirs, storage_dir, union_branch,
+    splash_dir = os.path.join(storage_dir_, "splash")
+    if os.path.exists(splash_dir):
+        src_ostree_archive_dir = os.path.join(storage_dir, "ostree-archive")
+        ub.create_splash_initramfs(splash_dir, src_ostree_archive_dir)
```
From what I can gather, this implementation is introducing a new approach regarding the management of the tool state, right? I mean, today all stateful commands create or update "changes" directories which are ready to be consumed by OSTree via the `union` command. With this new approach, we need to preprocess some intermediate directory in order to produce the changes directory before OSTree can consume it.

This approach looks sensible at first sight and it's probably better in terms of performance if we had followed it with all commands. Unfortunately that's not the case and introducing a new approach for a single command raises concerns about the impact it could have on both new and existing commands.

The new approach would basically force new commands operating in artifacts directly or indirectly produced from the intermediate ones to follow the same approach, that is, to postpone their work to when `union` is run. This means, errors related to them might also be detected later (during `union`) which would likely become a source of confusion to users of the stateful workflow.

Thinking a bit about existing commands:

Today we already have the command that signs the kernel and we are about to change the `splash` command to add support for FIT images. The kernel signing command takes the kernel as input and produces a new, signed one. With the new approach, we'd need to move the signing to the `union` command too (because only then we'd create a new ramdisk and use it to create a new kernel FIT image). But that would be strange to users, as the signing wouldn't happen during the `secboot sign-kernel` invocation but later. We can imagine cases where the user is keeping their keys in a PKCS#11 device and they'd expect the keys to be accessed during that invocation, which would make things even stranger.

@jsrc27 Given the above, I believe introducing this new approach at this point is not a good idea for us or for users of TorizonCore Builder. But, let me know your thoughts.

Anyway, since we're working with the `splash` command for FIT, we should probably wait a bit before continuing the work here to avoid hard-to-solve conflicts...

jsrc27 replied:
Well for secure boot/FIT images I did not take any of this into consideration as it was not in the scope of the ticket I was working on here.

In general, the only reason I went with this approach was because now we would have two commands that optionally customize the initramfs, that can be ran by the user in any order. Currently the `splash` command creates the customized initramfs file to be consumed later. If this approach were to be followed for `splash-config` there would be a weird ordering issue. Meaning whichever command was ran second would need to check if there was already a customization done, then recreate the customized initramfs file again to include the additional changes from whatever command was ran second.

It could be done, but for obvious reasons this approach seemed inefficient and redundant. Therefore I decided to move the creation of the initramfs to later in the process. At a point where TCB can know that all splash related commands have been run (or not run). That way any customizations can be considered at the same, meaning the initramfs only needs to be generated once if needed. It also avoids the issue of `splash` and `splash-config`  needing to "know" or check whether the other command has been ran first . Since ultimately these commands should be independent of one another.

> The new approach would basically force new commands operating in artifacts directly or indirectly produced from the intermediate ones to follow the same approach, that is, to postpone their work to when union is run. This means, errors related to them might also be detected later (during union) which would likely become a source of confusion to users of the stateful workflow.

Not sure if I understand the concern here. We only have splash screen related operations that create a customized initramfs currently. Currently we only have the one command that does so, meaning this command dictated things. Are we not allowed to change the precedent set by this one command that was created years ago?

rborn-tx replied:
Yes, the performance benefits are clear as I mentioned.

But, how did you intend to solve the issues I described above (which come from the fact that with secure boot images the ramdisk is inside the kernel FIT image)? I'm talking about this paragraph:

> Today we already have the command that signs the kernel and we are about to change the splash command to add support for FIT images. The kernel signing command takes the kernel as input and produces a new, signed one. With the new approach, we'd need to move the signing to the union command too (because only then we'd create a new ramdisk and use it to create a new kernel FIT image). But that would be strange to users, as the signing wouldn't happen during the secboot sign-kernel invocation but later. We can imagine cases where the user is keeping their keys in a PKCS#11 device and they'd expect the keys to be accessed during that invocation, which would make things even stranger.

You could think that this is out of scope because relates to Secure Boot, but if we don't think ahead, we could introduce a change that breaks the Secure Boot workflow which would lead to complicated discussions in the near future (notice that the support for the `splash` command with FIT-based images is in progress right now).

lucas-akira replied:
I haven't thought about the creation of the updated initramfs on the `union` command affecting the re-signing workflow. I think both points on keeping/re-working the changes are valid concerns:

- We could keep things in the PR as they are now, but that would require later moving the kernel signing step to the `union` command as well, which I agree that it would be strange, as `secboot sign-kernel` wouldn't actually sign anything.

- We could recreate the entire initramfs binary each time the splash/splash-config command is invoked, each time checking if a customization to initramfs was made beforehand, but that wouldn't be very efficient.

From what I understand currently for the `splash` command TCB creates a zipped CPIO file that only has the PNG image then appends it to what is effectively a copy of the unmodified initramfs file, which is also a zipped CPIO file. The end result is that the provided PNG image ends up overwriting the existing `watermark.png`.

Since `splash` and `splash-config` modify different files present in initramfs, their appending steps shouldn't interfere with one another, right?

If we remove the entire initramfs creation step in `splash` and `splash-config` and only append their additions to an existing initramfs file in the changes directory, wouldn't that solve the ordering issue? The unmodified initramfs binary could be copied to the changes directory during `images unpack`, so `splash`/`splash-config` would always have it available from the start.

jsrc27 replied:
> The unmodified initramfs binary could be copied to the changes directory during images unpack, so splash/splash-config would always have it available from the start.

So we'd always have an initramfs file in changes from the start no matter what. What if no initramfs customizations were done? Would `union` just "commit" this into the the new OSTree despite it not actually changing? I guess in theory for the user this wouldn't matter, though from a technical perspective it would be weird to commit something that didn't really change.

I guess `union` could check for the existence of other artifacts to see if an initramfs customization was done (i.e if `usr/share/plymouth/themes/spinner/` exists). Then, `union` could decide whether to include the initramfs in the commit or not. Though this creates an annoying precedent for future commands that may customize initramfs. As `union` would need to check for some indication if any of these commands were ran prior as well. In which case do we just commit the initramfs found in changes regardless no matter what every time?

Furthermore, another less than favorable thing is that each initramfs customizing command would still need to create their own zipped CPIO file to be appended. Rather than just doing it once for all initramfs customizations.

rborn-tx reviewed (commented):
I've stopped my review preemptively because the concern I've raised will likely change a lot the implementation.

jsrc27 committed:
splashconfig: Add new command related to customizing splash screen config

The new "splash-config" command can set a new Plymouth configuration
file. The command can also output the current active configuration.

Related-to: TCB-785

Signed-off-by: Jeremias Cordoba <jeremias.cordoba@toradex.com>

jsrc27 committed:
build: Add support for the new splash-config command to the build command

Related-to: TCB-785

Signed-off-by: Jeremias Cordoba <jeremias.cordoba@toradex.com>

jsrc27 committed:
tests: Add new tests for the splash-config command

Related-to: TCB-785

Signed-off-by: Jeremias Cordoba <jeremias.cordoba@toradex.com>

jsrc27 commented:
Rebased with the latest changes.

rborn-tx commented on tcbuilder/cli/splashconfig.py:
```diff
@@ -0,0 +1,148 @@
+        config_current = splashconfig_be.get_default_config()
+        log.info("Found default configuration file.")
+
+    create_config(args.output_file, config_current)
```
Would it make sense to factor this out into a `splash_config_dump()` function?

That would keep us aligned with the pattern we've been using, where `do_xxx()` focuses on connecting argument parsing (`ArgParse`) with the internal (CLI-level) API. It also makes reuse easier and clarifies what data the API actually needs, given that CLI arguments are wrapped in an args object with non-obvious properties.

jsrc27 replied:
Done in V2

rborn-tx commented on tcbuilder/cli/splashconfig.py:
```diff
@@ -0,0 +1,148 @@
+        help=("File path to the new configuration file."))
+    subparser.set_defaults(func=do_splash_config_set)
+
+    # splash-config dump
+    subparser = subparsers.add_parser(
+        "dump",
+        help=("Prints the current Plymouth configuration."),
+        allow_abbrev=False)
+    subparser.add_argument(
+        "--file",
+        dest="output_file",
+        help=("Writes current Plymouth configuration to the provided file name."),
+        required=False,
+        default=None)
+    subparser.set_defaults(func=do_splash_config_dump)
```
I just thought: Would it be too much to have a `--force` flag here to cause the output file to be overwritten if it existed?

jsrc27 replied:
Done in V2.

rborn-tx reviewed (commented):
Looks great! I've added some suggestions/questions nonetheless.

Btw, I see that the new implementation of `splash` made the addition of this new command much simpler compared to the previous attempt. What do you think?

lucas-akira commented on tcbuilder/cli/splashconfig.py:
```diff
@@ -0,0 +1,148 @@
+def create_config(output_file, config):
+    """Prints or writes the current active Plymouth Configuration file.
+
+    :param config_file: Filename for where to write the conent. If 'None'
```
I guess the parameter here should now be `output_file`.

jsrc27 replied:
Nice catch, done in V2.

lucas-akira reviewed (approved):
Code looks good to me. I only have one comment.

rborn-tx commented:
@jsrc27 Once this is merged, please remember to cherry-pick it to the new `trixie` branch.

rborn-tx reviewed (approved):
LGTM

rborn-tx commented:
@jsrc27 I've already created a PR updating trixie with the changes in this PR: https://github.com/torizon/torizoncore-builder/pull/124.
